File: backend/app/schemas/analytics.py
from datetime import datetime
from pydantic import BaseModel


class AnalyticsOverviewOut(BaseModel):
    total_students: int
    total_instructors: int
    total_organizations: int
    total_courses: int
    total_classes: int
    total_meetings: int
    closed_meetings: int
    certificates_issued: int
    active_subscriptions: int
    pending_charges_cents: int
    paid_charges_cents: int
    document_count: int
    attendance_rate: int
    completion_rate: int
    engagement_rate: int


class CourseAnalyticsOut(BaseModel):
    course_id: int
    course_name: str
    modality: str
    enrollments: int
    class_offerings: int
    meetings: int
    closed_meetings: int
    certificates_issued: int
    attendance_rate: int
    completion_rate: int
    average_best_quiz_score: int
    pending_charges_cents: int
    paid_charges_cents: int


class StudentAnalyticsOut(BaseModel):
    student_id: int
    student_name: str
    enrollments: int
    completed_courses: int
    certificates_issued: int
    attendance_rate: int
    progress_rate: int
    quiz_rate: int
    documents_count: int
    active_subscriptions: int


class ClassAnalyticsOut(BaseModel):
    class_offering_id: int
    class_name: str
    course_id: int
    course_name: str
    total_enrolled: int
    meetings: int
    closed_meetings: int
    present: int
    late: int
    absent: int
    attendance_rate: int
    waitlist_count: int
    certificates_issued: int


class CompletionReportRowOut(BaseModel):
    scope_type: str
    scope_id: int
    course_id: int
    course_name: str
    class_name: str | None = None
    enrolled: int
    completed: int
    completion_rate: int


class AttendanceReportRowOut(BaseModel):
    class_offering_id: int
    class_name: str
    course_id: int
    course_name: str
    meetings: int
    closed_meetings: int
    present: int
    late: int
    absent: int
    attendance_rate: int


class EngagementReportRowOut(BaseModel):
    course_id: int
    course_name: str
    progress_records: int
    completed_progress_records: int
    engagement_rate: int
    quiz_attempts: int
    passed_quiz_attempts: int
    quiz_pass_rate: int


class ClassPerformanceReportRowOut(BaseModel):
    class_offering_id: int
    class_name: str
    course_id: int
    course_name: str
    enrolled: int
    completion_rate: int
    attendance_rate: int
    average_best_quiz_score: int


class RoiReportRowOut(BaseModel):
    organization_id: int | None
    organization_name: str
    students: int
    paid_charges_cents: int
    pending_charges_cents: int
    certificates_issued: int
    completion_rate: int
    revenue_per_certificate_cents: int


class AnalyticsCourseSummaryOut(BaseModel):
    generated_at: datetime
    overview: AnalyticsOverviewOut
    courses: list[CourseAnalyticsOut]

File: frontend/src/types/analytics.ts
export interface AnalyticsOverview {
  total_students: number;
  total_instructors: number;
  total_organizations: number;
  total_courses: number;
  total_classes: number;
  total_meetings: number;
  closed_meetings: number;
  certificates_issued: number;
  active_subscriptions: number;
  pending_charges_cents: number;
  paid_charges_cents: number;
  document_count: number;
  attendance_rate: number;
  completion_rate: number;
  engagement_rate: number;
}

export interface CourseAnalytics {
  course_id: number;
  course_name: string;
  modality: string;
  enrollments: number;
  class_offerings: number;
  meetings: number;
  closed_meetings: number;
  certificates_issued: number;
  attendance_rate: number;
  completion_rate: number;
  average_best_quiz_score: number;
  pending_charges_cents: number;
  paid_charges_cents: number;
}

export interface StudentAnalytics {
  student_id: number;
  student_name: string;
  enrollments: number;
  completed_courses: number;
  certificates_issued: number;
  attendance_rate: number;
  progress_rate: number;
  quiz_rate: number;
  documents_count: number;
  active_subscriptions: number;
}

export interface ClassAnalytics {
  class_offering_id: number;
  class_name: string;
  course_id: number;
  course_name: string;
  total_enrolled: number;
  meetings: number;
  closed_meetings: number;
  present: number;
  late: number;
  absent: number;
  attendance_rate: number;
  waitlist_count: number;
  certificates_issued: number;
}

export interface CompletionReportRow {
  scope_type: 'course' | 'class';
  scope_id: number;
  course_id: number;
  course_name: string;
  class_name: string | null;
  enrolled: number;
  completed: number;
  completion_rate: number;
}

export interface AttendanceReportRow {
  class_offering_id: number;
  class_name: string;
  course_id: number;
  course_name: string;
  meetings: number;
  closed_meetings: number;
  present: number;
  late: number;
  absent: number;
  attendance_rate: number;
}

export interface EngagementReportRow {
  course_id: number;
  course_name: string;
  progress_records: number;
  completed_progress_records: number;
  engagement_rate: number;
  quiz_attempts: number;
  passed_quiz_attempts: number;
  quiz_pass_rate: number;
}

export interface ClassPerformanceReportRow {
  class_offering_id: number;
  class_name: string;
  course_id: number;
  course_name: string;
  enrolled: number;
  completion_rate: number;
  attendance_rate: number;
  average_best_quiz_score: number;
}

export interface RoiReportRow {
  organization_id: number | null;
  organization_name: string;
  students: number;
  paid_charges_cents: number;
  pending_charges_cents: number;
  certificates_issued: number;
  completion_rate: number;
  revenue_per_certificate_cents: number;
}
